### PR TITLE
added code modifications for 'database' scenario

### PR DESCRIPTION
**Pull request by deepchoudhery**
making code modifications for both `--type sqlserver-efcore` and `--type npgsql-efcore` scenarios.
- making relevant `Program.cs` changes
- adding a `NewDbContext` file to the root of the project and using this new DbContext in above changes.
- added `DbContextHelper` to hold related constants, dicts, and helpers for creating the NewDbContext.
  - `CreateDbContext` uses `TemplateFoldersUtilities` to get the relevant T4 file, and then uses our implementation of `ITemplateInvoker` to create and run the T4 template.
- `DbContextProperties` is used to format the relevant Program.cs code changes.
- added `NewDbContext.tt`, the T4 template for the new DbContext.
- added `db-apphost.json` and `db-webapi.json` for the new `Program.cs` code modifications.

### CODE REVIEW

File: src/Shared/Microsoft.DotNet.Scaffolding.Helpers/Roslyn/DocumentBuilder.cs [hunk 1]
```diff
@@ -846,29 +847,68 @@ internal static SyntaxNode AddExpressionToParent(SyntaxNode originalMethod, Code
...
+            }
 
-        if (modifiedExprNode is null)
-        {
-            return originalMethod;
+            // Replace existing expression with updated expression
+            updatedParent = parent.ReplaceNode(exprNode, modifiedExprNode);
         }
+        //add the scenario to check for an InvocationExpressionSyntax and update it if needed.
+        else if(invocationExpression != null)
```
Comment (tlmii) [inline src/Shared/Microsoft.DotNet.Scaffolding.Helpers/Roslyn/DocumentBuilder.cs:878]: nit: space after if

File: tools/dotnet-scaffold-aspire/Commands/DatabaseCommand.cs [hunk 2]
```diff
@@ -95,5 +159,176 @@ internal void InstallPackages(DatabaseCommandSettings commandSettings)
                     includePrerelease: commandSettings.Prerelease);
             }
         }
+
+        internal async Task<bool> UpdateAppHostAsync(DatabaseCommandSettings commandSettings)
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply tools/dotnet-scaffold-aspire/Commands/DatabaseCommand.cs:163]: yep internal was for testing purposes really, will change these all to be private (plus other commands) until we get unit tests.

File: src/Shared/Microsoft.DotNet.Scaffolding.Helpers/Roslyn/DocumentBuilder.cs (hunk 1 above)
Comment (deepchoudhery) [reply src/Shared/Microsoft.DotNet.Scaffolding.Helpers/Roslyn/DocumentBuilder.cs:878]: fixed ty

File: src/Shared/Microsoft.DotNet.Scaffolding.Helpers/Microsoft.DotNet.Scaffolding.Helpers.csproj [hunk 3]
```diff
@@ -19,14 +19,28 @@
...
 
+  <ItemGroup>
+    <None Update="Templates\DbContext\NewDbContext.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <LastGenOutput>NewDbContext.cs</LastGenOutput>
+    </None>
+  </ItemGroup>
+
+  <ItemGroup>
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/Shared/Microsoft.DotNet.Scaffolding.Helpers/Microsoft.DotNet.Scaffolding.Helpers.csproj:36]: don't think its needed, removed it. It gets added when the T4 template gets generated during build.

File: tools/dotnet-scaffold-aspire/Commands/DatabaseCommand.cs [hunk 4]
```diff
@@ -1,35 +1,65 @@
...
+                return 0;
+            }
+            else
+            {
+                _logger.LogMessage("An error occurred.");
+                return -1;
+            }
         }
 
         public class DatabaseCommandSettings : CommandSettings
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply tools/dotnet-scaffold-aspire/Commands/DatabaseCommand.cs:65]: moved to the bottom, tyty

File: tools/dotnet-scaffold-aspire/Commands/DatabaseCommand.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply tools/dotnet-scaffold-aspire/Commands/DatabaseCommand.cs:163]: all changed to internal

File: src/Shared/Microsoft.DotNet.Scaffolding.Helpers/T4Templating/TemplateFolderUtilities.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using System.Reflection;
+
+namespace Microsoft.DotNet.Scaffolding.Helpers.T4Templating;
+public static class TemplateFoldersUtilities
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/Shared/Microsoft.DotNet.Scaffolding.Helpers/T4Templating/TemplateFolderUtilities.cs:6]: ok creating `ITemplateFolderService`

File: src/Shared/Microsoft.DotNet.Scaffolding.Helpers/T4Templating/TemplateFolderUtilities.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+{
+    public static List<string> GetTemplateFolders(string[] baseFolders)
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/Shared/Microsoft.DotNet.Scaffolding.Helpers/T4Templating/TemplateFolderUtilities.cs:8]: sounds good

File: tools/dotnet-scaffold-aspire/Commands/GetCommands.cs [hunk 7]
```diff
@@ -53,7 +54,7 @@ public static class GetCmdsHelper
         internal static Parameter ProjectParameter = new() { Name = "--project", DisplayName = "Web or worker project file", Description = "Web or worker project associated with the Aspire App host", Required = true, Type = BaseTypes.String, PickerType = InteractivePickerType.ProjectPicker };
 
         internal static List<string> CachingTypeCustomValues = ["redis", "redis-with-output-caching"];
-        internal static List<string> DatabaseTypeCustomValues = ["npgsql", "npgsql-efcore", "sqlserver-efcore", "cosmos-efcore"];
+        internal static List<string> DatabaseTypeCustomValues = DbContextHelper.DatabaseTypes;
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply tools/dotnet-scaffold-aspire/Commands/GetCommands.cs:57]: I was thinking just with extensibility, other aspire components/tools might want this list of supported database types. But yes, the supported database types will differ based on the tool/library.

[the remark this answers is not in the document]
Comment (deepchoudhery) [reply tools/dotnet-scaffold-aspire/Commands/GetCommands.cs:57]: ok fair fair, will revert this change

**Review by tlmii — APPROVED** [review]

**Pull request merged**